File: Assignment3/functional_requirements.py
import pandas as pd

class FunctionalRequirements:
    def __init__(self, service):
        self.add_service = service

    #  Generic filter function
    def _apply_filters(self, students, gender=None, age=None, class_id=None, city=None, pincode=None):
        filtered = students

        if gender:
            filtered = [s for s in filtered if s.gender == gender]
        if age:
            filtered = [s for s in filtered if s.age == age]
        if class_id:
            filtered = [s for s in filtered if s.class_id == class_id]

        if city:
            student_ids = [addr.student_id for addr in self.add_service.addresses if addr.city == city]
            filtered = [s for s in filtered if s.id in student_ids]

        if pincode:
            student_ids = [addr.student_id for addr in self.add_service.addresses if addr.pincode == pincode]
            filtered = [s for s in filtered if s.id in student_ids]

        return filtered

    #  1. Find Students by Pincode
    def students_by_pincode(self, pincode, **filters):
        students = [s for s in self.add_service.students
                    if any(s.id == addr.student_id for addr in self.add_service.addresses if addr.pincode == pincode)]
        students = self._apply_filters(students, **filters)
        return pd.DataFrame([s.model_dump() for s in students]) if students else pd.DataFrame()

    #  2. Find Students by City
    def students_by_city(self, city, **filters):
        students = [s for s in self.add_service.students
                    if any(s.id == addr.student_id for addr in self.add_service.addresses if addr.city == city)]
        students = self._apply_filters(students, **filters)
        return pd.DataFrame([s.model_dump() for s in students]) if students else pd.DataFrame()

    #  3. Find Students by Class
    def students_by_class(self, class_id, **filters):
        students = [s for s in self.add_service.students if s.class_id == class_id]
        students = self._apply_filters(students, **filters)
        return pd.DataFrame([s.model_dump() for s in students]) if students else pd.DataFrame()

    #  4. Get Passed Students
    def passed_students(self, passing_marks=50, **filters):
        students = [s for s in self.add_service.students if s.marks >= passing_marks]
        students = self._apply_filters(students, **filters)
        return pd.DataFrame([s.model_dump() for s in students]) if students else pd.DataFrame()

    #  5. Get Failed Students
    def failed_students(self, passing_marks=50, **filters):
        students = [s for s in self.add_service.students if s.marks < passing_marks]
        students = self._apply_filters(students, **filters)
        return pd.DataFrame([s.model_dump() for s in students]) if students else pd.DataFrame()

    #  6. Ranking
    def ranking(self, **filters):
        students = sorted(self.add_service.students, key=lambda s: s.marks, reverse=True)
        students = self._apply_filters(students, **filters)
        return pd.DataFrame([s.model_dump() for s in students]) if students else pd.DataFrame()

File: Assignment3/Java/services/FunctionalRequirements.java
package Assignment3.Java.services;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

import Assignment3.Java.models.Address;
import Assignment3.Java.models.Student;

public class FunctionalRequirements {
    private final Service service;

    public FunctionalRequirements(Service service) {
        this.service = service;
    }

    private List<Student> applyFilters(List<Student> students, String gender, Integer age, Integer classId, String city, String pincode) {
        List<Student> filtered = new ArrayList<>(students);
        if (gender != null) {
            filtered = filtered.stream().filter(s -> gender.equals(s.getGender())).collect(Collectors.toList());
        }
        if (age != null) {
            filtered = filtered.stream().filter(s -> s.getAge() == age).collect(Collectors.toList());
        }
        if (classId != null) {
            filtered = filtered.stream().filter(s -> s.getClassId() == classId).collect(Collectors.toList());
        }
        if (city != null || pincode != null) {
            List<Integer> studentIds = service.getAddresses().stream()
                    .filter(a -> (city == null || city.equals(a.getCity())) && (pincode == null || pincode.equals(a.getPincode())))
                    .map(Address::getStudentId)
                    .collect(Collectors.toList());
            filtered = filtered.stream().filter(s -> studentIds.contains(s.getId())).collect(Collectors.toList());
        }
        return filtered;
    }

    public List<Student> topNStudents(int n, String gender, Integer classId) {
        List<Student> students = new ArrayList<>(service.getStudents());
        if (gender != null || classId != null) {
            students = applyFilters(students, gender, null, classId, null, null);
        }
        return students.stream().sorted(Comparator.comparingDouble(Student::getMarks).reversed()).limit(n).collect(Collectors.toList());
    }

    public List<Student> studentsByCity(String city) {
        List<Integer> ids = service.getAddresses().stream().filter(a -> city.equals(a.getCity())).map(Address::getStudentId).collect(Collectors.toList());
        return service.getStudents().stream().filter(s -> ids.contains(s.getId())).collect(Collectors.toList());
    }

    public List<Student> failedStudents(double passingMarks, String gender, Integer classId) {
        List<Student> students = service.getStudents().stream().filter(s -> s.getMarks() < passingMarks).collect(Collectors.toList());
        return applyFilters(students, gender, null, classId, null, null);
    }

    public List<Student> ranking(String gender, Integer classId) {
        List<Student> students = new ArrayList<>(service.getStudents());
        students = applyFilters(students, gender, null, classId, null, null);
        students.sort(Comparator.comparingDouble(Student::getMarks).reversed());
        return students;
    }
}
